c/g6-read-iterator.c

```diff
@@ -0,0 +1,795 @@
+int _decodeGraph(char *graphBuff, const int graphOrder, const int numChars, graphP pGraph)
+{
+	int exitCode = OK;
+
+	if (pGraph == NULL)
+	{
+		ErrorMessage("Must initialize graph datastructure before invoking _decodeGraph.\n");
+		return NOTOK;
+	}
+
+	int numPaddingZeroes = _getExpectedNumPaddingZeroes(graphOrder, numChars);
+
+	char currByte = '\0';
+	int bitValue = 0;
+	int row = 0;
+	int col = 1;
+	for (int i = 0; i < numChars; i++)
+	{
+		currByte = graphBuff[i] - 63;
+		// j corresponds to the number of places one must bitshift the byte by to read
+		// the next bit in the byte
+		for (int j = sizeof(char) * 5; j >= 0; j--)
+		{
+			// If we are on the final byte, we know that the final numPaddingZeroes bits
+			// can be ignored, so we break out of the loop
+			if ((i == numChars) && j == numPaddingZeroes - 1)
+				break;
+
+			if (row == col)
+			{
+				row = 0;
+				col++;
+			}
+
+			bitValue = ((currByte >> j) & 1u) ? 1 : 0;
+			if (bitValue == 1)
+			{
+				// Add gp_GetFirstVertex(pGraph), which is 1 if NIL == 0 (i.e. 1-based labelling) and 0 if NIL == -1 (0-based)
+				exitCode = gp_AddEdge(pGraph, row+gp_GetFirstVertex(pGraph), 0, col+gp_GetFirstVertex(pGraph), 0);
+				// TODO: verify if breaking out entirely from decode is acceptable on gp_AddEdge returning NOTOK
+				if (exitCode == NOTOK)
+					return exitCode;
+			}
+
+			row++;
+		}    
+	}
+
+	pGraph->N = graphOrder;
+
+	return exitCode;
+}
+
+int endG6ReadIteration(G6ReadIterator *pG6ReadIterator)
+{
+	int exitCode = OK;
+
+	if (pG6ReadIterator != NULL)
+	{
+		if (pG6ReadIterator->g6Infile != NULL && pG6ReadIterator->doIOwnFilePointer)
+		{
+			int fcloseCode = fclose(pG6ReadIterator->g6Infile);
+
+			if (fcloseCode != 0)
+			{
+				ErrorMessage("Unable to close G6ReadIterator's g6Infile.\n");
+				exitCode = NOTOK;
+			}
+		}
+
+		pG6ReadIterator->g6Infile = NULL;
+
+		if (pG6ReadIterator->currGraphBuff != NULL)
+		{
+			free(pG6ReadIterator->currGraphBuff);
+			pG6ReadIterator->currGraphBuff = NULL;
+		}
+	}
+
+	return exitCode;
+}
+
+int freeG6ReadIterator(G6ReadIterator **ppG6ReadIterator)
+{
+	int exitCode = OK;
+
+	if (ppG6ReadIterator != NULL && (*ppG6ReadIterator) != NULL)
+	{
+		if ((*ppG6ReadIterator)->g6Infile != NULL && (*ppG6ReadIterator)->doIOwnFilePointer)
+		{
+			int fcloseCode = fclose((*ppG6ReadIterator)->g6Infile);
+
+			if (fcloseCode != 0)
+			{
+				ErrorMessage("Unable to close G6ReadIterator's g6Infile.\n");
+				exitCode = NOTOK;
+			}
+		}
+
+		(*ppG6ReadIterator)->g6Infile = NULL;
+
+		(*ppG6ReadIterator)->numGraphsRead = 0;
+		(*ppG6ReadIterator)->graphOrder = 0;
+
+		if ((*ppG6ReadIterator)->currGraphBuff != NULL)
+		{
+			free((*ppG6ReadIterator)->currGraphBuff);
+			(*ppG6ReadIterator)->currGraphBuff = NULL;
+		}
+
+		(*ppG6ReadIterator)->currGraph = NULL;
+
+		free((*ppG6ReadIterator));
+		(*ppG6ReadIterator) = NULL;
+	}
+
+	return exitCode;
+}
+
+int _ReadGraphFromG6String(graphP pGraphToRead, char *g6EncodedString)
```

Comment:
I've now added the `strOrFile` container, and have refactored `_ReadGraphFromG6String` to actually read from string (instead of using `fputs` to dump the string into a `tmpfile` and calling `_ReadGraphFromG6FilePointer`).